### PR TITLE
TASK-1325177: Certificates to leverage ESO

### PR DESCRIPTION
**Pull request by pega-Abhinav**

### CODE REVIEW

**Comment by kishorv10** [comment]
@pega-Abhinav  Is [this](https://github.com/pegasystems/pega-helm-charts/blob/0d9387739ada0ee798a63ad8a61fd72e77cb0128/charts/pega/templates/_pega-deployment.tpl#L63) condition should have OR check for `certificateSecrets`? I doubt if certificates are mounting from secrets if certifcates field is empty.

File: charts/pega/templates/_helpers.tpl [hunk 1]
```diff
@@ -13,11 +13,19 @@
...
+    sources:
+  {{ if (.Values.global.certificatesSecret) }}
+  {{- range .Values.global.certificatesSecret }}
+    - secret:
+        name: {{ . }}
+  {{- end }}
+  {{ else }}
+    # This name will be referred in the volume mounts kind.
+    - secret:
+        name: {{ template "pegaImportCertificatesConfig" $ }}
```
Comment (kishorv10) [inline charts/pega/templates/_helpers.tpl:27]: I recommend merging `charts/pega/templates/_pega-certificates-config.tpl` code to this place, this way multiple conditions can be avoided and `_pega-certificates-config.tpl` can be removed altogether. 
Or do vice-versa

File: charts/pega/values.yaml [hunk 2]
```diff
@@ -14,7 +14,10 @@ global:
   actions:
     execute: "deploy"
 
-  # Add custom certificates to be mounted to container, as a map
+  # Add custom certificates to be mounted to container
+  #to support external secrets
+  #supports multiple externalSecrets , add externalSecret names as array of strings
+  certificatesSecret: []
```
[the remark this answers is not in the document]
Comment (pega-Abhinav) [reply charts/pega/values.yaml:20]: Sure , will update

File: charts/pega/templates/_helpers.tpl (hunk 1 above)
Comment (pega-Abhinav) [reply charts/pega/templates/_helpers.tpl]: Discussed with @kishorv10  on this , hence resolving

**Review by kishorv10 — APPROVED** [review]

File: charts/pega/README.md [hunk 3]
```diff
@@ -880,13 +880,13 @@ To support this option,
 3) Update your local Helm repository to the latest version using the command:
     - helm repo update pega https://pegasystems.github.io/pega-helm-charts
 4) Update your values.yaml file to refer to the external secret manager for certificates.
-5) pega supports multiple externalSecrets for custom certificates, add externalSecret names as array of strings to certificatesSecret
+5) pega supports multiple externalSecrets for custom certificates, add externalSecret names as array of strings to certificatesSecrets
```
Comment (taz-pega-work) [inline charts/pega/README.md:883]: 5) Add multiple custom certificates that you maintain as an externally-managed secret, each as a string, separated by a comma in the `certificatesSecrets`  parameter.

File: charts/pega/README.md [hunk 4]
```diff
@@ -880,13 +880,13 @@ To support this option,
...
 
 • You can either pass certificates as external secrets or as plain text to values.yaml ,not both. if both are given only external secrets will be mounted to into the tomcat container
```
Comment (taz-pega-work) [inline charts/pega/README.md:885]: update:
You can either pass certificates as external secrets or as plain text to the Pega values.yaml, but not both. If you provide both, the deployment mounts only external secrets into the tomcat container.

do you want to add something about how it's not secure if you expose certificates in plain text?? this sounds like an accident waiting to happen ;-)

**Review by taz-pega-work — COMMENTED** [review]
Can you please make a few wording adjustments?

File: charts/pega/values-large.yaml [hunk 5]
```diff
@@ -15,8 +15,8 @@ global:
   actions:
     execute: "deploy"
   # Add custom certificates to be mounted to container
-  #to support external secrets
-  #supports multiple externalSecrets , add externalSecret names as array of strings
+  # to support external secrets
+  # supports multiple externalSecrets , add externalSecret names as array of strings
```
Comment (taz-pega-work) [inline charts/pega/values-large.yaml:19]: update these across the 3 values.yaml files>
 to support an external secret, specify your secret using the `certificates` parameter.
 to support multiple external secrets, specify each of your external secrets as an array of comma-separated strings using the `certificatesSecrets` parameter.

File: charts/pega/values.yaml [hunk 6]
```diff
@@ -15,8 +15,8 @@ global:
     execute: "deploy"
 
   # Add custom certificates to be mounted to container
-  #to support external secrets
-  #supports multiple externalSecrets , add externalSecret names as array of strings
+  # to support external secrets
```
Comment (taz-pega-work) [inline charts/pega/values.yaml:18]: to support an external secret, specify your secret using the certificates parameter.
to support multiple external secrets, specify each of your external secrets as an array of comma-separated strings using the certificatesSecrets parameter.

**Review by taz-pega-work — COMMENTED** [review]
a few small changes to clarify which parameter gets which certificate strings.

File: charts/pega/README.md (hunk 3 above)
Comment (pega-Abhinav) [reply charts/pega/README.md]: Fixed this

File: charts/pega/README.md (hunk 4 above)
Comment (pega-Abhinav) [reply charts/pega/README.md]: Fixed this

File: charts/pega/values-large.yaml (hunk 5 above)
Comment (pega-Abhinav) [reply charts/pega/values-large.yaml]: here we use certificates parameter to provide list of certificates as plain text.
Where as certificatesSecrets takes names of external secrets as array of strings.

Updated accordingly in the latest commit , please review

**Review by Saurabh-16 — APPROVED** [review]

File: charts/pega/values-large.yaml [hunk 7]
```diff
@@ -16,7 +16,7 @@ global:
     execute: "deploy"
   # Add custom certificates to be mounted to container
   # to support custom certificates as plain text, pass certificates to certificates parameter
-  # to support multiple custom certificates as external secrets, specify each of your external secrets as an array of comma-separated strings using the certificatesSecrets parameter.
```
Comment (taz-pega-work) [inline charts/pega/values-large.yaml]: the change is fine if you really have to save space, but I don't think it's clear using certs = certificates...whatever you decide, please make the terms uniform for all of the options...and passing certificates directly isn't a good idea, is it? Don't we say somewhere else to encrypt the certificate if you aren't using secrets? I'd remind them here to encrypt.

Add custom certs to be mounted to the container
to support custom certs as plain text (less secure), pass them directly to certificates parameter;
to support multiple custom certs as external secrets, specify each of your external secrets as an array of comma-separated strings using the certificatesSecrets parameter.

**Review by taz-pega-work — COMMENTED** [review]
do we really have to shorten certificates to certs? I don't think it looks professional. If you drop it here, you have to drop it in the other sections in the values.yaml files, since consistency is the main goal.

File: charts/pega/values-large.yaml (hunk 7 above)
Comment (pega-Abhinav) [reply charts/pega/values-large.yaml]: Reduced certificates to certs as the line length was exceeding 180 characters , which is not allowed.

Can you please suggest some alternative for below sentence , with reduced characters.. ? will update accordingly
"to support multiple custom certificates as external secrets, specify each of your external secrets as an array of comma-separated strings using the certificatesSecrets parameter."

Comment (pega-Abhinav) [reply charts/pega/values-large.yaml]: Fixed with below changes

# Add custom certificates to be mounted to the container
# to support custom certificates as plain text (less secure), pass them directly using the certificates parameter;
# to support multiple custom certificates as external secrets, specify each of your external secrets
# as an array of comma-separated strings using the certificatesSecrets parameter.

**Review by taz-pega-work — APPROVED** [review]
thanks for incorporating my review comments

**Pull request merged**